### PR TITLE
panic: add __attribute__((noreturn)) #if __GNUC__

### PR DESCRIPTION
**Pull request by marc-hb**
Commit 9e1a2b251a3e9d added ((analyzer_noreturn)) but only for clang.

Fixes at least these two -Werr build failures and probably similar
others with a different .config or in the future:
```
   sof/src/probe/probe.c: In function 'probe_cb_produce':
   sof/src/probe/probe.c:635:11: error: 'valid_bytes' and
   'container_bytes' may be used uninitialized in this function
   [-Werror=maybe-uninitialized]
```
Build failure reproduced with xtensa-apl-elf-gcc (crosstool-NG
1.23.0.440-3b8c) 7.3.0

### CODE REVIEW

File: tools/testbench/panic.c [hunk 1]
```diff
@@ -6,4 +6,8 @@
 
 #include <sof/debug/panic.h>
 
-void __panic(uint32_t p, char *filename, uint32_t linenum) { }
+void __panic(uint32_t p, char *filename, uint32_t linenum)
+{
+	while (1)
+		;
+}
```
Comment (marc-hb) [inline tools/testbench/panic.c]: I'm not familiar with the testbench so not sure what to do here. @jajanusz would  some `while(1) sleep` be better?

Comment (jajanusz) [reply tools/testbench/panic.c]: something like `abort()` will be better, cos at least program will crash and it's easier to debug than freeze

Comment (jajanusz) [reply tools/testbench/panic.c]: Also tools should be more clever when we switch eventually to c11, cos there is `_Noreturn void abort(void);`

File: src/include/sof/debug/panic.h [hunk 2]
```diff
@@ -12,20 +12,26 @@
 #include <ipc/trace.h>
 #include <stdint.h>
 
+#ifdef SOF_NO_RET
+#error "SOF_NO_RET already defined!?"
```
Comment (jajanusz) [inline src/include/sof/debug/panic.h]: No need for this case, compiler will complain anyway if it's redefined

File: src/include/sof/debug/panic.h [hunk 3]
```diff
@@ -12,20 +12,26 @@
 #include <ipc/trace.h>
 #include <stdint.h>
 
+#ifdef SOF_NO_RET
```
Comment (jajanusz) [inline src/include/sof/debug/panic.h]: I'd prefer more verbose name like, `SOF_NO_RETURN`

File: tools/testbench/panic.c (hunk 1 above)
Comment (jajanusz) [reply tools/testbench/panic.c]: @marc-hb It should be separate commit - before adding attributes

**Comment by marc-hb** [comment]
tl;dr: this is blocked on https://github.com/thesofproject/cmocka/pull/4

I reproduced the "Internal Intel CI" failure which goes like this (thanks @jajanusz !)
```
06:37:57,210 WARN  - cc1: warnings being treated as errors
06:37:57,210 WARN  - /quickbuild/workspace1/24733/SOF_FW/test/cmocka/src/common_mocks.c: In function ‘__panic’:
06:37:57,210 WARN  - /quickbuild/workspace1/24733/SOF_FW/test/cmocka/src/common_mocks.c:73: warning: ‘noreturn’ function does return
06:37:57,211 WARN  - make[2]: *** [test/cmocka/CMakeFiles/common_mock.dir/src/common_mocks.c.o] Error 2
```
The https://sof-ci.01.org/sof-pr-viewer/#/build/PR2581/build4863969 system apparently doesn't collect or share `stderr`... yet, @jajanusz (again! :-) is looking into that. 

Fixing this -Werror requires a small, `noreturn` cmocka.h change submitted here: https://github.com/thesofproject/cmocka/pull/4

**Comment by jajanusz** [comment]
https://github.com/thesofproject/cmocka/pull/4 was merged, you can continue working on this, thanks

File: tools/testbench/panic.c (hunk 1 above)
Comment (jajanusz) [reply tools/testbench/panic.c]: Github shows that it's 2nd commit. I preferred it to be the 1st, cos It may break git bisect for testbench if you have noreturn attribute for function without abort in it, but I'm not sure.

Comment (marc-hb) [reply tools/testbench/panic.c]: The addition of the `abort()` is the first commit. It's hard to tell because commits are ordered sometimes chronologically and sometimes in reverse order and because 
[github doesn't care about git history](https://github.com/isaacs/github/issues/959#issuecomment-593131237) (vendor lock-in), but if you click on this link and check the "parent" you can make sure: https://github.com/thesofproject/sof/commit/a343042b959dfbb

Comment (marc-hb) [reply tools/testbench/panic.c]: > Also tools should be more clever when we switch eventually to c11,

https://stackoverflow.com/questions/48615184/does-visual-studio-2017-fully-support-c99

(no)

**Comment by marc-hb** [comment]
- Summary of https://sof-ci.01.org/sofpr/PR2581/build5550/devicetest

The entire check-sof-logger line fails with the usual:
```
error: abi version in /etc/sof/sof-cml.ldc file does not coincide with abi version used by logger.
error: logger ABI Version is 4:2:0
error: ldc_file ABI Version is 3:14:0
```
I believe this is tracked in #2507. Could we temporarily disable this line until #2507 is fixed?

The bottom half of APL_UP2_PCM512X failed with
`arecord: set_params:1363: Unable to install hw params:` or `Unable to set hw params for playback: Invalid argument` or `Setting of hwparams failed: Invalid argument`. 
This can't be related to this PR...

- Summary of https://sof-ci.01.org/sof-pr-viewer/#/build/PR2581/build4892406
While appearing as successful,  it was actually interrupted by a timeout long before completing most tests.

Also found this error in some TGL log, saw it in different and unrelated PR too (#2626)
```
ninja bin -v
Command return code: 1
Command error output: CMake Warning at CMakeLists.txt:79 (message):
Couldn't get compiler version
```
The timeout and the compiler version error don't seem related to each other.

File: tools/testbench/panic.c (hunk 1 above)
Comment (jajanusz) [reply tools/testbench/panic.c]: You're right, I forgot that GH doesn't care about order and sorts by time ...

**Comment by jajanusz** [comment]
> Also found this error in some TGL log, saw it in different and unrelated PR too (#2626)
> 
> ```
> ninja bin -v
> Command return code: 1
> Command error output: CMake Warning at CMakeLists.txt:79 (message):
> Couldn't get compiler version
> ```
> 
> The timeout and the compiler version error don't seem related to each other.
> 
> EDIT: I found a comparison with https://sof-ci.01.org/sof-pr-viewer/#/build/PR2626/build4891579 interesting. It has the same (TGL win?) cmake error and the same timeout internally, however the sof-pr-viewer output looks completely different.

TGL for windows step is broken, that's why it's marked as optional, safe to ignore.
